### PR TITLE
update config_api with initialization and recovery utilities

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 config API doesn't initialize config data, doesn't provide utilities for config data recovery

 #### Summary of Changes
 * require Default of config data
 * add utility
 * update test

Fixes #

### CODE REVIEW

File: programs/config_api/src/config_instruction.rs [hunk 1]
```diff
@@ -1,27 +1,15 @@
...
-use crate::ConfigState;
-use bincode::serialize;
-use serde_derive::{Deserialize, Serialize};
-use solana_sdk::instruction::{AccountMeta, Instruction};
-use solana_sdk::pubkey::Pubkey;
-use solana_sdk::short_vec;
-use solana_sdk::system_instruction;
+use crate::{ConfigKeys, ConfigState};
+use solana_sdk::{
+    instruction::{AccountMeta, Instruction},
```
[the remark this answers is not in the document]
Comment (rob-solana) [reply programs/config_api/src/config_instruction.rs]: I stole it from @mark-solana

File: programs/config_api/src/config_instruction.rs [hunk 2]
```diff
@@ -1,27 +1,15 @@
...
-
-impl ConfigKeys {
-    pub fn serialized_size(keys: Vec<(Pubkey, bool)>) -> usize {
-        serialize(&ConfigKeys { keys })
-            .unwrap_or_else(|_| vec![])
-            .len()
-    }
+fn initialize_account<T: ConfigState>(config_pubkey: &Pubkey) -> Instruction {
+    let account_metas = vec![AccountMeta::new(*config_pubkey, true)];
+    let account_data = (ConfigKeys { keys: vec![] }, T::default());
```
Comment (rob-solana) [inline programs/config_api/src/config_instruction.rs]: @mvines @CriesofCarrots this line is suspect for me, I think we're basically equivalent to before, but perhaps we should actually be putting the keys in the account at create()+init() time.

the reason I didn't just use create()+store() was that a fair few places where create() was being called weren't signing according to the metas produced by they store() logic...

File: programs/config_api/src/config_instruction.rs [hunk 3]
```diff
@@ -30,15 +18,18 @@ pub fn create_account<T: ConfigState>(
...
+    let space = T::max_space() + ConfigKeys::serialized_size(keys);
+    vec![
+        system_instruction::create_account(
+            from_account_pubkey,
+            config_account_pubkey,
+            lamports,
+            space,
+            &id(),
+        ),
+        initialize_account::<T>(config_account_pubkey),
```
Comment (CriesofCarrots) [inline programs/config_api/src/config_instruction.rs]: Just trying to understand what problem this solves. Are you perhaps adding a new config usage that has non-zero defaults?

Comment (rob-solana) [reply programs/config_api/src/config_instruction.rs]: here, I was mostly trying to fix a test ;)

I don't know if we can assume that a buffer of all zeros is always safely deserialize-able, and wanted for config data (after create()) to always be deserialize-able

File: programs/config_api/src/config_processor.rs [hunk 4]
```diff
@@ -100,380 +100,4 @@ pub fn process_instruction(
...
-        solana_logger::setup();
-        let (bank, mint_keypair) = create_bank(10_000);
-        let (bank_client, config_keypair) = create_config_account(bank, &mint_keypair, vec![]);
-        let config_account_data = bank_client
-            .get_account_data(&config_keypair.pubkey())
-            .unwrap()
-            .unwrap();
-        assert_eq!(
-            MyConfig::default(),
-            MyConfig::deserialize(&config_account_data).unwrap()
```
Comment (rob-solana) [inline programs/config_api/src/config_processor.rs]: this line was incorrect, testing the wrong thing

File: programs/config_tests/tests/config_processor.rs [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    solana_logger::setup();
+    let (bank, mint_keypair) = create_bank(10_000);
+    let (bank_client, config_keypair) = create_config_account(bank, &mint_keypair, vec![]);
+    let config_account_data = bank_client
+        .get_account_data(&config_keypair.pubkey())
+        .unwrap()
+        .unwrap();
+    assert_eq!(
+        Some(MyConfig::default()),
+        deserialize(get_config_data(&config_account_data).unwrap()).ok()
```
Comment (rob-solana) [inline programs/config_tests/tests/config_processor.rs]: the "fix", verified by MyConfig::default() being non-zero

File: programs/config_api/src/config_instruction.rs (hunk 3 above)
Comment (CriesofCarrots) [reply programs/config_api/src/config_instruction.rs]: Ah, okay. A fair precaution, and I could envision the non-zero default being useful at some point

File: programs/config_api/src/config_instruction.rs (hunk 2 above)
Comment (CriesofCarrots) [reply programs/config_api/src/config_instruction.rs]: Example?
Not certain I follow

**Pull request merged**